taichi/transforms/auto_diff.cpp

```diff
@@ -365,19 +567,34 @@ class MakeAdjoint : public IRVisitor {
   }
 
   void visit(RangeForStmt *for_stmt) override {
-    if (for_depth > 0)  // reverse non-parallelized for-loops
-      for_stmt->reverse();
-    for_depth += 1;
-    alloca_block = for_stmt->body.get();
-    for_stmt->body->accept(this);
-    for_depth -= 1;
+    auto new_for = for_stmt->clone();
+    auto new_for_ptr = (RangeForStmt *)new_for.get();
+    new_for_ptr->reversed = !new_for_ptr->reversed;
+    insert_back(std::move(new_for));
+    int len = new_for_ptr->body->statements.size();
+
+    for (int i = 0; i < len; i++) {
+      new_for_ptr->body->erase(0);
+    }
+
+    std::vector<Stmt *> statements;
+    // always make a copy since the list can be modified.
+    for (auto &stmt : for_stmt->body->statements) {
+      statements.push_back(stmt.get());
+    }
+    std::reverse(statements.begin(), statements.end());  // reverse-mode AD...
+    auto old_alloca_block = alloca_block;
+    for (auto stmt : statements) {
+      alloca_block = new_for_ptr->body.get();
+      current_block = new_for_ptr->body.get();
+      stmt->accept(this);
+    }
+    alloca_block = old_alloca_block;
   }
 
   void visit(StructForStmt *for_stmt) override {
```

Comment:
Will we encounter any `StructForStmt`s here?

Reply:
Probably no. I'll just be conservative here and leave this function here. Let's remove it in the future when we have more time.